### PR TITLE
safeString2 returns second value if first value is an empty string ""

### PR DESCRIPTION
**Pull request by samgermain**
2022-05-27T01:31:49.506Z
Node.js: v14.17.0
CCXT v1.84.5
PHP v8.1.6

`safe_string`, `safe_string_lower`, `safe_string_upper`, `safe_number` all just check for `test`

```                                                                                         
JS                                        PYTHON

{"test":"","test2":"10"}                  {'test': '', 'test2': '10'}
safe_string: undefined                    safe_string: None
safe_string_lower: undefined              safe_string_lower: None
safe_string_upper: undefined              safe_string_upper: None
safe_string_2: 10                         safe_string_2: 10
safe_string_lower_2: 10                   safe_string_lower_2: 10
safe_string_upper_2: 10                   safe_string_upper_2: 10
safe_number: undefined                    safe_number: None
safe_number_2: 10                         safe_number_2: 10.0
                                          
{"test":"5","test2":""}                   {'test': '5', 'test2': ''}
safe_string: 5                            safe_string: 5
safe_string_lower: 5                      safe_string_lower: 5
safe_string_upper: 5                      safe_string_upper: 5
safe_string_2: 5                          safe_string_2: 5
safe_string_lower_2: 5                    safe_string_lower_2: 5
safe_string_upper_2: 5                    safe_string_upper_2: 5
safe_number: 5                            safe_number: 5.0
safe_number_2: 5                          safe_number_2: 5.0
                                          
{"test":"5","test2":"10"}                 {'test': '5', 'test2': '10'}
safe_string: 5                            safe_string: 5
safe_string_lower: 5                      safe_string_lower: 5
safe_string_upper: 5                      safe_string_upper: 5
safe_string_2: 5                          safe_string_2: 5
safe_string_lower_2: 5                    safe_string_lower_2: 5
safe_string_upper_2: 5                    safe_string_upper_2: 5
safe_number: 5                            safe_number: 5.0
safe_number_2: 5                          safe_number_2: 5.0
                                          
{"test":5}                                {'test': 5}
safe_string: 5                            safe_string: 5
safe_string_lower: 5                      safe_string_lower: 5
safe_string_upper: 5                      safe_string_upper: 5
safe_string_2: 5                          safe_string_2: 5
safe_string_lower_2: 5                    safe_string_lower_2: 5
safe_string_upper_2: 5                    safe_string_upper_2: 5
safe_number: 5                            safe_number: 5.0
safe_number_2: 5                          safe_number_2: 5.0
                                          
{"test2":"5"}                             {'test2': '5'}
safe_string: undefined                    safe_string: None
safe_string_lower: undefined              safe_string_lower: None
safe_string_upper: undefined              safe_string_upper: None
safe_string_2: 5                          safe_string_2: 5
safe_string_lower_2: 5                    safe_string_lower_2: 5
safe_string_upper_2: 5                    safe_string_upper_2: 5
safe_number: undefined                    safe_number: None
safe_number_2: 5                          safe_number_2: 5.0
                                          
{"test":""}                               {'test': ''}
safe_string: undefined                    safe_string: None
safe_string_lower: undefined              safe_string_lower: None
safe_string_upper: undefined              safe_string_upper: None
safe_string_2: undefined                  safe_string_2: None
safe_string_lower_2: undefined            safe_string_lower_2: None
safe_string_upper_2: undefined            safe_string_upper_2: None
safe_number: undefined                    safe_number: None
safe_number_2: undefined                  safe_number_2: None
                                          
{"test2":""}                              {'test2': ''}
safe_string: undefined                    safe_string: None
safe_string_lower: undefined              safe_string_lower: None
safe_string_upper: undefined              safe_string_upper: None
safe_string_2: undefined                  safe_string_2: None
safe_string_lower_2: undefined            safe_string_lower_2: None
safe_string_upper_2: undefined            safe_string_upper_2: None
safe_number: undefined                    safe_number: None
safe_number_2: undefined                  safe_number_2: None
```

### CODE REVIEW

**Comment by kroitor** [comment]
@samgermain what about Python?

**Comment by samgermain** [comment]
> @samgermain what about Python?

There's nothing to change for python, it's not a problem

**Comment by samgermain** [comment]
@kroitor I couldn't get my php tests working, I made this script

```
<?php

$root = dirname(dirname(dirname(__FILE__)));

include $root . '/ccxt.php';

$exchange = new \ccxt\Exchange();

$example_1 = array(
    "test" => "",
    "test2" => "10",
);
$example_2 = array(
    "test" => "5",
    "test2" => "",
);
$example_2 = array(
    "test" => "5",
    "test2" => "10",
);
$example_3 = array(
    "test" => "5",
);
$example_4 = array(
    "test2" => "5",
);
$example_5 = array(
    "test" => "",
);
$example_6 = array(
    "test2" => "",
);
$examples = array(
    $example_1,
    $example_2,
    $example_2,
    $example_3,
    $example_4,
    $example_5,
    $example_6,
);

foreach ($examples as $ex) {
    echo "";
    echo json_encode($ex);
    echo $exchange->safe_string($ex, 'test');
    echo $exchange->safe_string_lower($ex, 'test');
    echo $exchange->safe_string_upper($ex, 'test');
    echo $exchange->safe_string_2($ex, 'test', 'test2');
    echo $exchange->safe_string_lower_2($ex, 'test', 'test2');
    echo $exchange->safe_string_upper_2($ex, 'test', 'test2');
    echo $exchange->safe_string_n($ex, 'test');
    echo $exchange->safe_string_lower_n($ex, 'test', 'test2');
    echo $exchange->safe_string_upper_n($ex, 'test', 'test2');
    echo $exchange->safe_number($ex, 'test');
    echo $exchange->safe_number_2($ex, 'test', 'test2');
}

?>
```

but am getting 

```
{"test":"","test2":"10"}PHP Fatal error:  Uncaught Error: Call to undefined function safe_string() in /path/ccxt/examples/php/test_safe_methods.php:46
Stack trace:
#0 {main}
  thrown in /path/ccxt/examples/php/test_safe_methods.php on line 46

Fatal error: Uncaught Error: Call to undefined function safe_string() in /path/ccxt/examples/php/test_safe_methods.php:46
Stack trace:
#0 {main}
  thrown in /path/ccxt/examples/php/test_safe_methods.php on line 46
```

**Comment by kroitor** [comment]
When I execute the php tests on my side, they're ok, but they're still at their very early stage, they don't test much really. The base safe-method functionality in PHP is not tested (still have to write the tests for that). I've ran your script on my side though with the following results (I will add some fixes to it shortly):

```PHP
~/ccxt $ php -f examples/php/test-safe-string.php
{"test":"","test2":"10"}PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test200{"test":"5","test2":"10"}555555PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test255{"test":"5","test2":"10"}555555PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test255{"test":"5"}555555PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test255{"test2":"5"}555PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test25{"test":""}PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test200{"test2":""}PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test2PHP Warning:  Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546

Warning: Invalid argument supplied for foreach() in /Users/igorkroitor/ccxt/php/Exchange.php on line 546
test20%
~/ccxt $
```

**Comment by kroitor** [comment]
PHP requires explicit newlines btw:

```PHP
<?php

$root = dirname(dirname(dirname(__FILE__)));

include $root . '/ccxt.php';

$exchange = new \ccxt\Exchange();

$example_1 = array(
    "test" => "",
    "test2" => "10",
);
$example_2 = array(
    "test" => "5",
    "test2" => "",
);
$example_2 = array(
    "test" => "5",
    "test2" => "10",
);
$example_3 = array(
    "test" => "5",
);
$example_4 = array(
    "test2" => "5",
);
$example_5 = array(
    "test" => "",
);
$example_6 = array(
    "test2" => "",
);
$examples = array(
    $example_1,
    $example_2,
    $example_2,
    $example_3,
    $example_4,
    $example_5,
    $example_6,
);

foreach ($examples as $ex) {
    echo "" . "\n";
    echo json_encode($ex) . "\n";
    echo $exchange->safe_string($ex, 'test') . "\n";
    echo $exchange->safe_string_lower($ex, 'test') . "\n";
    echo $exchange->safe_string_upper($ex, 'test') . "\n";
    echo $exchange->safe_string_2($ex, 'test', 'test2') . "\n";
    echo $exchange->safe_string_lower_2($ex, 'test', 'test2') . "\n";
    echo $exchange->safe_string_upper_2($ex, 'test', 'test2') . "\n";
    echo $exchange->safe_string_n($ex, 'test') . "\n";
    echo $exchange->safe_string_lower_n($ex, 'test', 'test2') . "\n";
    echo $exchange->safe_string_upper_n($ex, 'test', 'test2') . "\n";
    echo $exchange->safe_number($ex, 'test') . "\n";
    echo $exchange->safe_number_2($ex, 'test', 'test2') . "\n";
}

?>
```

**Comment by kroitor** [comment]
This version of your script worked with this PR, but I found a bug in the PHP implementation of safe_string_n, will add some fixes:

```PHP
<?php

$root = dirname(dirname(dirname(__FILE__)));

include $root . '/ccxt.php';

$exchange = new \ccxt\Exchange();

$example_1 = array(
    "test" => "",
    "test2" => "10",
);
$example_2 = array(
    "test" => "5",
    "test2" => "",
);
$example_2 = array(
    "test" => "5",
    "test2" => "10",
);
$example_3 = array(
    "test" => "5",
);
$example_4 = array(
    "test2" => "5",
);
$example_5 = array(
    "test" => "",
);
$example_6 = array(
    "test2" => "",
);
$examples = array(
    $example_1,
    $example_2,
    $example_2,
    $example_3,
    $example_4,
    $example_5,
    $example_6,
);

foreach ($examples as $ex) {
    echo "" . "\n";
    echo 'json_encode($ex) ' . json_encode($ex) . "\n";
    echo 'safe_string($ex, \'test\') ' . var_export($exchange->safe_string($ex, 'test'), true) . "\n";
    echo 'safe_string_lower($ex, \'test\') ' . var_export($exchange->safe_string_lower($ex, 'test'), true) . "\n";
    echo 'safe_string_upper($ex, \'test\') ' . var_export($exchange->safe_string_upper($ex, 'test'), true) . "\n";
    echo 'safe_string_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_2($ex, 'test', 'test2'), true) . "\n";
    echo 'safe_string_lower_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_lower_2($ex, 'test', 'test2'), true) . "\n";
    echo 'safe_string_upper_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_upper_2($ex, 'test', 'test2'), true) . "\n";
    echo 'safe_string_n($ex, array(\'test\')) ' . var_export($exchange->safe_string_n($ex, array('test')), true) . "\n";
    echo 'safe_string_lower_n($ex, array(\'test\', \'test2\')) ' . var_export($exchange->safe_string_lower_n($ex, array('test', 'test2')), true) . "\n";
    echo 'safe_string_upper_n($ex, array(\'test\', \'test2\')) ' . var_export($exchange->safe_string_upper_n($ex, array('test', 'test2')), true) . "\n";
    echo 'safe_number($ex, \'test\') ' . var_export($exchange->safe_number($ex, 'test'), true) . "\n";
    echo 'safe_number_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_number_2($ex, 'test', 'test2'), true) . "\n";
}

?>
```

-------------

```PHP
~/ccxt $ php -f examples/php/test-safe-string.php

json_encode($ex) {"test":"","test2":"10"}
safe_string($ex, 'test') NULL
safe_string_lower($ex, 'test') NULL
safe_string_upper($ex, 'test') NULL
safe_string_2($ex, 'test', 'test2') '10'
safe_string_lower_2($ex, 'test', 'test2') '10'
safe_string_upper_2($ex, 'test', 'test2') '10'
safe_string_n($ex, array('test')) NULL
safe_string_lower_n($ex, array('test', 'test2')) NULL <<<<<<<<<<<<<<<< should be 10
safe_string_upper_n($ex, array('test', 'test2')) NULL
safe_number($ex, 'test') NULL
safe_number_2($ex, 'test', 'test2') 10.0

json_encode($ex) {"test":"5","test2":"10"}
safe_string($ex, 'test') '5'
safe_string_lower($ex, 'test') '5'
safe_string_upper($ex, 'test') '5'
safe_string_2($ex, 'test', 'test2') '5'
safe_string_lower_2($ex, 'test', 'test2') '5'
safe_string_upper_2($ex, 'test', 'test2') '5'
safe_string_n($ex, array('test')) '5'
safe_string_lower_n($ex, array('test', 'test2')) '5'
safe_string_upper_n($ex, array('test', 'test2')) '5'
safe_number($ex, 'test') 5.0
safe_number_2($ex, 'test', 'test2') 5.0

json_encode($ex) {"test":"5","test2":"10"}
safe_string($ex, 'test') '5'
safe_string_lower($ex, 'test') '5'
safe_string_upper($ex, 'test') '5'
safe_string_2($ex, 'test', 'test2') '5'
safe_string_lower_2($ex, 'test', 'test2') '5'
safe_string_upper_2($ex, 'test', 'test2') '5'
safe_string_n($ex, array('test')) '5'
safe_string_lower_n($ex, array('test', 'test2')) '5'
safe_string_upper_n($ex, array('test', 'test2')) '5'
safe_number($ex, 'test') 5.0
safe_number_2($ex, 'test', 'test2') 5.0

json_encode($ex) {"test":"5"}
safe_string($ex, 'test') '5'
safe_string_lower($ex, 'test') '5'
safe_string_upper($ex, 'test') '5'
safe_string_2($ex, 'test', 'test2') '5'
safe_string_lower_2($ex, 'test', 'test2') '5'
safe_string_upper_2($ex, 'test', 'test2') '5'
safe_string_n($ex, array('test')) '5'
safe_string_lower_n($ex, array('test', 'test2')) '5'
safe_string_upper_n($ex, array('test', 'test2')) '5'
safe_number($ex, 'test') 5.0
safe_number_2($ex, 'test', 'test2') 5.0

json_encode($ex) {"test2":"5"}
safe_string($ex, 'test') NULL
safe_string_lower($ex, 'test') NULL
safe_string_upper($ex, 'test') NULL
safe_string_2($ex, 'test', 'test2') '5'
safe_string_lower_2($ex, 'test', 'test2') '5'
safe_string_upper_2($ex, 'test', 'test2') '5'
safe_string_n($ex, array('test')) NULL
safe_string_lower_n($ex, array('test', 'test2')) '5'
safe_string_upper_n($ex, array('test', 'test2')) '5'
safe_number($ex, 'test') NULL
safe_number_2($ex, 'test', 'test2') 5.0

json_encode($ex) {"test":""}
safe_string($ex, 'test') NULL
safe_string_lower($ex, 'test') NULL
safe_string_upper($ex, 'test') NULL
safe_string_2($ex, 'test', 'test2') NULL
safe_string_lower_2($ex, 'test', 'test2') NULL
safe_string_upper_2($ex, 'test', 'test2') NULL
safe_string_n($ex, array('test')) NULL
safe_string_lower_n($ex, array('test', 'test2')) NULL
safe_string_upper_n($ex, array('test', 'test2')) NULL
safe_number($ex, 'test') NULL
safe_number_2($ex, 'test', 'test2') NULL

json_encode($ex) {"test2":""}
safe_string($ex, 'test') NULL
safe_string_lower($ex, 'test') NULL
safe_string_upper($ex, 'test') NULL
safe_string_2($ex, 'test', 'test2') NULL
safe_string_lower_2($ex, 'test', 'test2') NULL
safe_string_upper_2($ex, 'test', 'test2') NULL
safe_string_n($ex, array('test')) NULL
safe_string_lower_n($ex, array('test', 'test2')) NULL
safe_string_upper_n($ex, array('test', 'test2')) NULL
safe_number($ex, 'test') NULL
safe_number_2($ex, 'test', 'test2') NULL
~/ccxt $
```

File: js/hitbtc3.js [hunk 1]
```diff
@@ -530,9 +530,9 @@ module.exports = class hitbtc3 extends Exchange {
             let withdrawEnabled = undefined;
             for (let j = 0; j < rawNetworks.length; j++) {
                 const rawNetwork = rawNetworks[j];
-                let networkId = this.safeString (rawNetwork, 'protocol', '');
-                if (networkId.length === 0) {
-                    networkId = this.safeString (rawNetwork, 'network', '');
+                let networkId = this.safeString (rawNetwork, 'protocol');
+                if (networkId !== undefined) {
+                    networkId = this.safeString (rawNetwork, 'network');
```
Comment (samgermain) [inline js/hitbtc3.js:535]: This should be `=== undefined`, but it should just be `safeString2` I'll update it

**Comment by samgermain** [comment]
> This version of your script worked with this PR, but I found a bug in the PHP implementation of safe_string_n, will add some fixes:
> 
> ```
> <?php
> 
> $root = dirname(dirname(dirname(__FILE__)));
> 
> include $root . '/ccxt.php';
> 
> $exchange = new \ccxt\Exchange();
> 
> $example_1 = array(
>     "test" => "",
>     "test2" => "10",
> );
> $example_2 = array(
>     "test" => "5",
>     "test2" => "",
> );
> $example_2 = array(
>     "test" => "5",
>     "test2" => "10",
> );
> $example_3 = array(
>     "test" => "5",
> );
> $example_4 = array(
>     "test2" => "5",
> );
> $example_5 = array(
>     "test" => "",
> );
> $example_6 = array(
>     "test2" => "",
> );
> $examples = array(
>     $example_1,
>     $example_2,
>     $example_2,
>     $example_3,
>     $example_4,
>     $example_5,
>     $example_6,
> );
> 
> foreach ($examples as $ex) {
>     echo "" . "\n";
>     echo 'json_encode($ex) ' . json_encode($ex) . "\n";
>     echo 'safe_string($ex, \'test\') ' . var_export($exchange->safe_string($ex, 'test'), true) . "\n";
>     echo 'safe_string_lower($ex, \'test\') ' . var_export($exchange->safe_string_lower($ex, 'test'), true) . "\n";
>     echo 'safe_string_upper($ex, \'test\') ' . var_export($exchange->safe_string_upper($ex, 'test'), true) . "\n";
>     echo 'safe_string_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_2($ex, 'test', 'test2'), true) . "\n";
>     echo 'safe_string_lower_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_lower_2($ex, 'test', 'test2'), true) . "\n";
>     echo 'safe_string_upper_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_string_upper_2($ex, 'test', 'test2'), true) . "\n";
>     echo 'safe_string_n($ex, array(\'test\')) ' . var_export($exchange->safe_string_n($ex, array('test')), true) . "\n";
>     echo 'safe_string_lower_n($ex, array(\'test\', \'test2\')) ' . var_export($exchange->safe_string_lower_n($ex, array('test', 'test2')), true) . "\n";
>     echo 'safe_string_upper_n($ex, array(\'test\', \'test2\')) ' . var_export($exchange->safe_string_upper_n($ex, array('test', 'test2')), true) . "\n";
>     echo 'safe_number($ex, \'test\') ' . var_export($exchange->safe_number($ex, 'test'), true) . "\n";
>     echo 'safe_number_2($ex, \'test\', \'test2\') ' . var_export($exchange->safe_number_2($ex, 'test', 'test2'), true) . "\n";
> }
> 
> ?>
> ```
> 
> ```
> ~/ccxt $ php -f examples/php/test-safe-string.php
> 
> json_encode($ex) {"test":"","test2":"10"}
> safe_string($ex, 'test') NULL
> safe_string_lower($ex, 'test') NULL
> safe_string_upper($ex, 'test') NULL
> safe_string_2($ex, 'test', 'test2') '10'
> safe_string_lower_2($ex, 'test', 'test2') '10'
> safe_string_upper_2($ex, 'test', 'test2') '10'
> safe_string_n($ex, array('test')) NULL
> safe_string_lower_n($ex, array('test', 'test2')) NULL <<<<<<<<<<<<<<<< should be 10
> safe_string_upper_n($ex, array('test', 'test2')) NULL
> safe_number($ex, 'test') NULL
> safe_number_2($ex, 'test', 'test2') 10.0
> 
> json_encode($ex) {"test":"5","test2":"10"}
> safe_string($ex, 'test') '5'
> safe_string_lower($ex, 'test') '5'
> safe_string_upper($ex, 'test') '5'
> safe_string_2($ex, 'test', 'test2') '5'
> safe_string_lower_2($ex, 'test', 'test2') '5'
> safe_string_upper_2($ex, 'test', 'test2') '5'
> safe_string_n($ex, array('test')) '5'
> safe_string_lower_n($ex, array('test', 'test2')) '5'
> safe_string_upper_n($ex, array('test', 'test2')) '5'
> safe_number($ex, 'test') 5.0
> safe_number_2($ex, 'test', 'test2') 5.0
> 
> json_encode($ex) {"test":"5","test2":"10"}
> safe_string($ex, 'test') '5'
> safe_string_lower($ex, 'test') '5'
> safe_string_upper($ex, 'test') '5'
> safe_string_2($ex, 'test', 'test2') '5'
> safe_string_lower_2($ex, 'test', 'test2') '5'
> safe_string_upper_2($ex, 'test', 'test2') '5'
> safe_string_n($ex, array('test')) '5'
> safe_string_lower_n($ex, array('test', 'test2')) '5'
> safe_string_upper_n($ex, array('test', 'test2')) '5'
> safe_number($ex, 'test') 5.0
> safe_number_2($ex, 'test', 'test2') 5.0
> 
> json_encode($ex) {"test":"5"}
> safe_string($ex, 'test') '5'
> safe_string_lower($ex, 'test') '5'
> safe_string_upper($ex, 'test') '5'
> safe_string_2($ex, 'test', 'test2') '5'
> safe_string_lower_2($ex, 'test', 'test2') '5'
> safe_string_upper_2($ex, 'test', 'test2') '5'
> safe_string_n($ex, array('test')) '5'
> safe_string_lower_n($ex, array('test', 'test2')) '5'
> safe_string_upper_n($ex, array('test', 'test2')) '5'
> safe_number($ex, 'test') 5.0
> safe_number_2($ex, 'test', 'test2') 5.0
> 
> json_encode($ex) {"test2":"5"}
> safe_string($ex, 'test') NULL
> safe_string_lower($ex, 'test') NULL
> safe_string_upper($ex, 'test') NULL
> safe_string_2($ex, 'test', 'test2') '5'
> safe_string_lower_2($ex, 'test', 'test2') '5'
> safe_string_upper_2($ex, 'test', 'test2') '5'
> safe_string_n($ex, array('test')) NULL
> safe_string_lower_n($ex, array('test', 'test2')) '5'
> safe_string_upper_n($ex, array('test', 'test2')) '5'
> safe_number($ex, 'test') NULL
> safe_number_2($ex, 'test', 'test2') 5.0
> 
> json_encode($ex) {"test":""}
> safe_string($ex, 'test') NULL
> safe_string_lower($ex, 'test') NULL
> safe_string_upper($ex, 'test') NULL
> safe_string_2($ex, 'test', 'test2') NULL
> safe_string_lower_2($ex, 'test', 'test2') NULL
> safe_string_upper_2($ex, 'test', 'test2') NULL
> safe_string_n($ex, array('test')) NULL
> safe_string_lower_n($ex, array('test', 'test2')) NULL
> safe_string_upper_n($ex, array('test', 'test2')) NULL
> safe_number($ex, 'test') NULL
> safe_number_2($ex, 'test', 'test2') NULL
> 
> json_encode($ex) {"test2":""}
> safe_string($ex, 'test') NULL
> safe_string_lower($ex, 'test') NULL
> safe_string_upper($ex, 'test') NULL
> safe_string_2($ex, 'test', 'test2') NULL
> safe_string_lower_2($ex, 'test', 'test2') NULL
> safe_string_upper_2($ex, 'test', 'test2') NULL
> safe_string_n($ex, array('test')) NULL
> safe_string_lower_n($ex, array('test', 'test2')) NULL
> safe_string_upper_n($ex, array('test', 'test2')) NULL
> safe_number($ex, 'test') NULL
> safe_number_2($ex, 'test', 'test2') NULL
> ~/ccxt $
> ```

I was having trouble with `safeStringN` in JavaScript also, but I am also having these problems on the `master` branch, so it's not related to changes in this PR

```
/Users/sam/Documents/dev/CCXT/ccxt/js/base/functions/type.js:19
    , getValueFromKeysInArray = (object, array) => object[array.find (k => prop (object,k) !== undefined)]
                                                                ^

TypeError: array.find is not a function
    at getValueFromKeysInArray (/Users/sam/Documents/dev/CCXT/ccxt/js/base/functions/type.js:19:65)
    at binance.safeStringN (/Users/sam/Documents/dev/CCXT/ccxt/js/base/functions/type.js:71:69)
    at /Users/sam/Documents/dev/CCXT/ccxt/examples/js/test_safe_methods.js:47:44
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/sam/Documents/dev/CCXT/ccxt/examples/js/test_safe_methods.js:38:3)
    at Module._compile (internal/modules/cjs/loader.js:1068:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
    at Module.load (internal/modules/cjs/loader.js:933:32)
    at Function.Module._load (internal/modules/cjs/loader.js:774:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
```

File: js/binance.js [hunk 2]
```diff
@@ -3660,7 +3660,7 @@ module.exports = class binance extends Exchange {
                 tag = undefined;
             }
         }
-        let txid = this.safeString (transaction, 'txId');
+        let txid = this.safeString (transaction, 'txId', '');
```
Comment (kroitor) [inline js/binance.js:3663]: I don't think that we should fill in these strings, if they're missing. They should not be empty.

Comment (kroitor) [reply js/binance.js:3663]: @samgermain can you revert the commit that added the edits in derived files in this PR?

Comment (samgermain) [reply js/binance.js:3663]: @kroitor All the ones that get a default of an empty string get it because we later call something like `.length` or `.indexOf` on it, so the program will crash if it's set to undefined,

File: js/bitfinex2.js [hunk 3]
```diff
@@ -1314,7 +1314,7 @@ module.exports = class bitfinex2 extends bitfinex {
             }
         }
         let status = undefined;
-        const statusString = this.safeString (order, 13);
+        const statusString = this.safeString (order, 13, '');
```
Comment (kroitor) [inline js/bitfinex2.js:1317]: This should not be an empty string either.

File: js/bitso.js [hunk 4]
```diff
@@ -880,7 +880,7 @@ module.exports = class bitso extends Exchange {
             'fund_currency': currency['id'],
         };
         const response = await this.privateGetFundingDestination (this.extend (request, params));
-        let address = this.safeString (response['payload'], 'account_identifier');
+        let address = this.safeString (response['payload'], 'account_identifier', '');
```
Comment (kroitor) [inline js/bitso.js:883]: The address cannot be an empty string. We're missing a proper parseDepositAddress here, and a checkAddress call, but I would leave that our for a standalone PR. Let's remove this default value.

File: js/digifinex.js [hunk 5]
```diff
@@ -723,7 +723,7 @@ module.exports = class digifinex extends Exchange {
         const id = this.safeString (trade, 'id');
         const orderId = this.safeString (trade, 'order_id');
         const timestamp = this.safeTimestamp2 (trade, 'date', 'timestamp');
-        let side = this.safeString2 (trade, 'type', 'side');
+        let side = this.safeString2 (trade, 'type', 'side', '');
```
Comment (kroitor) [inline js/digifinex.js:726]: The side cannot be empty here.

**Comment by samgermain** [comment]
@kroitor I thought we wanted to remove all the default empty string values

**Comment by kroitor** [comment]
> @kroitor I thought we wanted to remove all the default empty string values

I did remove some of them upon merging, yes.
